src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Node.cs

```diff
@@ -0,0 +1,160 @@
+        /// <param name="node">The <see cref="JsonNode"/> to convert.</param>
+        /// <param name="returnType">The type of the object to convert to and return.</param>
+        /// <param name="context">A metadata provider for serializable types.</param>
+        /// <exception cref="System.ArgumentNullException">
+        /// <paramref name="returnType"/> is <see langword="null"/>.
+        ///
+        /// -or-
+        ///
+        /// <paramref name="context"/> is <see langword="null"/>.
+        /// </exception>
+        /// <exception cref="JsonException">
+        /// The JSON is invalid.
+        ///
+        /// -or-
+        ///
+        /// <paramref name="returnType" /> is not compatible with the JSON.
+        ///
+        /// -or-
+        ///
+        /// There is remaining data in the string beyond a single JSON value.</exception>
+        /// <exception cref="NotSupportedException">
+        /// There is no compatible <see cref="System.Text.Json.Serialization.JsonConverter"/>
+        /// for <paramref name="returnType"/> or its serializable members.
+        /// </exception>
+        /// <exception cref="InvalidOperationException">
+        /// The <see cref="JsonSerializerContext.GetTypeInfo(Type)"/> method of the provided
+        /// <paramref name="context"/> returns <see langword="null"/> for the type to convert.
+        /// </exception>
+        public static object? Deserialize(this JsonNode? node, Type returnType, JsonSerializerContext context)
+        {
+            if (returnType == null)
+            {
+                throw new ArgumentNullException(nameof(returnType));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            return ReadUsingMetadata<object?>(node, GetTypeInfo(context, returnType));
+        }
+
+        [RequiresUnreferencedCode(SerializationUnreferencedCodeMessage)]
+        private static TValue? ReadUsingOptions<TValue>(JsonNode? node, Type returnType, JsonSerializerOptions? options) =>
+            ReadUsingMetadata<TValue>(node, GetTypeInfo(returnType, options));
+
+        private static TValue? ReadUsingMetadata<TValue>(JsonNode? node, JsonTypeInfo jsonTypeInfo)
+        {
+            JsonSerializerOptions options = jsonTypeInfo.Options;
+            Debug.Assert(options != null);
+
+            // For performance, we avoid calling JsonSerializer.Deserialize(node) since that will allocate an extra byte[].
+
+            using (var output = new PooledByteBufferWriter(options.DefaultBufferSize))
+            {
+                using (var writer = new Utf8JsonWriter(output, options.GetWriterOptions()))
+                {
+                    if (node is null)
+                    {
+                        writer.WriteNullValue();
+                    }
+                    else
+                    {
+                        node.WriteTo(writer, options);
```

Comment:
I wish there was a way we could avoid writing to an intermediate buffer. At this point it's probably not possible with `Utf8JsonReader` being a concrete ref struct.

Reply:
At least we only have one intermediate buffer; the current workarounds (before this new support) would have 2.

In theory, we could add another type of reader to avoid this.